### PR TITLE
ipq40xx:add support qxwlan e2600ac-c1/c2

### PR DESCRIPTION
**Pull request by zhangsd20**
Qxwlan E2600AC C1 based on IPQ4019

Specifications:
SOC:	Qualcomm IPQ4019
DRAM:	256 MiB
FLASH:	32 MiB Winbond W25Q256
ETH:	Qualcomm QCA8075
WLAN:	5G + 5G/2.4G
	* 2T2R 2.4/5 GHz
  	 - QCA4019 hw1.0 (SoC)
  	 - uses AP-DK07 BDF from QCA4019/hw1.0/board-2.bin
	* 2T2R 5 GHz
	 - QCA4019 hw1.0 (SoC)
	 - uses AP-DK07 BDF from QCA4019/hw1.0/board-2.bin
INPUT:  Reset buutton
LED:	1x Power ,6 driver by gpio
SERIAL: UART (J5)
UUSB:	USB3.0
POWER:	1x DC jack for main power input (9-24 V)
SLOT:	Pcie (J25), sim card (J11), SD card (J51)

Flash instruction (using U-Boot CLI and tftp server):

 - Configure PC with static IP 192.168.1.10 and tftp server.
 - Rename "sysupgrade" filename to "firmware.bin" and place it in tftp
   server directory.
 - Connect PC with one of RJ45 ports, power up the board and press
   "enter" key to access U-Boot CLI.
 - Use the following command to update the device to OpenWrt: "run lfw".

Flash instruction (using U-Boot web-based recovery):

 - Configure PC with static IP 192.168.1.xxx(2-254)/24.
 - Connect PC with one of RJ45 ports, press the reset button, power up
   the board and keep button pressed for around 6-7 seconds, until LEDs
   start flashing.
 - Open your browser and enter 192.168.1.1, select "sysupgrade" image
   and click the upgrade button.

Qxwlan E2600AC C2 based on IPQ4019

Specifications:
SOC:	Qualcomm IPQ4019
DRAM:	256 MiB
NOR:	16 MiB Winbond W25Q128
NAND:	128MiB Micron MT29F1G08ABAEAWP
ETH:	Qualcomm QCA8075
WLAN:	5G + 5G/2.4G
	* 2T2R 2.4/5 GHz
  	 - QCA4019 hw1.0 (SoC)
  	 - uses AP-DK07 BDF from QCA4019/hw1.0/board-2.bin
	* 2T2R 5 GHz
	 - QCA4019 hw1.0 (SoC)
	 - uses AP-DK07 BDF from QCA4019/hw1.0/board-2.bin
INPUT:  Reset buutton
LED:	1x Power ,6 driver by gpio
SERIAL: UART (J5)
USB:	USB3.0
POWER:	1x DC jack for main power input (9-24 V)
SLOT:	Pcie (J25), sim card (J11), SD card (J51)

Flash instruction (using U-Boot CLI and tftp server):

 - Configure PC with static IP 192.168.1.10 and tftp server.
 - Rename "ubi" filename to "ubi-firmware.bin" and place it in tftp
   server directory.
 - Connect PC with one of RJ45 ports, power up the board and press
   "enter" key to access U-Boot CLI.
 - Use the following command to update the device to OpenWrt: "run lfw".

Flash instruction (using U-Boot web-based recovery):

 - Configure PC with static IP 192.168.1.xxx(2-254)/24.
 - Connect PC with one of RJ45 ports, press the reset button, power up
   the board and keep button pressed for around 6-7 seconds, until LEDs
   start flashing.
 - Open your browser and enter 192.168.1.1, select "ubi" image
   and click the upgrade button.

Signed-off-by: 张鹏 <sd20@qxwlan.com>

### CODE REVIEW

**Comment by robimarko** [comment]
Where are custom BDF-s included?
I have seen that you submited them upstream for inclusion into board-2.bin but until they are merged you need to include them in OpenWrt by using the ipq-wifi package.

**Comment by zhangsd20** [comment]
What should I do if we use ap-dk07 board-2.bin?

**Comment by robimarko** [comment]
Well, what is qcom,ath10k-calibration-variant = "Qxwlan-E2600AC"; used for if you are using upstream ap-dk07 board-2.bin?

File: target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata [hunk 1]
```diff
@@ -138,7 +138,9 @@ case "$FIRMWARE" in
 	compex,wpj428 |\
 	engenius,eap1300 |\
 	openmesh,a42 |\
-	openmesh,a62)
+	openmesh,a62 \|
```
Comment (chunkeey) [inline target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata]: What happend here? Shouldn't this be "|\"?

**Comment by chunkeey** [comment]
The old thread for reference: https://github.com/openwrt/openwrt/pull/1655

> What should I do if we use ap-dk07 board-2.bin?

Wait a sec... This makes it sound like you are purposefully skipping out on doing the calibration on your own boards?! :disappointed:  Isn't this going to be a problem for your customers down the line or what am I missing here?

**Comment by pepe2k** [comment]
@chunkeey 

> Wait a sec... This makes it sound like you are purposefully skipping out on doing the calibration on your own boards?! disappointed Isn't this going to be a problem for your customers down the line or what am I missing here?

Assuming that the board strictly follows QCA reference design (QCA gives you not only schematics but basically everything, including PCB design, BOM wit recommended parts/vendors, etc.) for the RF part, it should work (I have seen *calibrated* IPQ4k based boards using QCA BDFs for all 3, 2 or just one radio and custom ones for the rest), see also: http://lists.infradead.org/pipermail/ath10k/2017-January/009104.html

> [...] Each board data is custom for the board layout / part selection - it's a template that is used during calibration. The data in OTP is just a diff against the board template  (board.bin / board-2.bin.) [...]

Cheers,
Piotr

File: target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata (hunk 1 above)
Comment (zhangsd20) [reply target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata]: i'm so sorry.It's my mistake.

**Comment by pepe2k** [comment]
@chunkeey 
> From what I know, the board.bin data did indeed serve as a template during calibration for the WAVE-1 QCA9880/2/7. In the thread Michael clarified: "[notably qca61x4 hw3+ and qca4019 need distinct board files to be uploaded. Otherwise they fail in various ways.](https://www.mail-archive.com/ath10k@lists.infradead.org/msg05893.html). And this was later confirmed by Rajkumar Manoharan later in the [post](https://www.mail-archive.com/ath10k@lists.infradead.org/msg06501.html) I linked in the #1655.
> 
> > In QCA4019 platform, only radio specific calibration (pre-cal-data) is stored in flash. Board specific contents are read from board-2.bin. For each radio appropriate board data should be loaded. To fetch correct board data from board-2.bin bundle, pre-cal/radio specific caldata should be loaded first to get proper board id.

My understanding here is that there could be two different IPQ40xx based designs using the very same board.bin for one (or more) WiFi radios, as far as designs of their RF sides are the same (same parts and components values used)*.

> What I do find very odd in regards to how the board-2.bin for the QCA4019 is handled by QCA and their partners.

I don't.
Different target markets, different internal teams taking care of the subject (?), plus... it's still Qualcomm: mess everywhere and with everything, wheels invented from scratch, datasheets for new ICs made with copy&paste approach with mistakes, etc.**

> Because for the customer QCA9377/QCA6174 device, Qualcomm takes much more care of the board-2.bin integration process. For example if you look into https://github.com/kvalo/ath10k-firmware/blob/master/QCA6174/hw3.0/board-2.bin you'll notice that there are many board definitions in there.

I think that market for QCA6174 is much bigger (phones, PCs running Windows, MiFi, etc.) and company focuses more resources on it than in case of IPQ40xx.

> They all have their unique identifiers (pciid + variant) and multiple boardfiles have the same checksum, so having a the data twice isn't a problem either.

Yeah but what's the purpose of having the very same binary data multiple times in that file? For me that's a resource wasting approach (why not assign single board.bin to several different pciid?) - anyone with QCA6174 card has to download almost 0.5 MB (for now, it will get bigger) binary just to make the card working.

*source/quotation needed
**this is only my private opinion

Cheers,
Piotr

**Comment by chunkeey** [comment]
> > From what I know, the board.bin data did indeed serve as a template during calibration for the WAVE-1 QCA9880/2/7. In the thread Michael clarified: "[notably qca61x4 hw3+ and qca4019 need distinct board files to be uploaded. Otherwise they fail in various ways.](https://www.mail-archive.com/ath10k@lists.infradead.org/msg05893.html). And this was later confirmed by Rajkumar Manoharan later in the [post](https://www.mail-archive.com/ath10k@lists.infradead.org/msg06501.html) I linked in the #1655.
> > > In QCA4019 platform, only radio specific calibration (pre-cal-data) is stored in flash. Board specific contents are read from board-2.bin. For each radio appropriate board data should be loaded. To fetch correct board data from board-2.bin bundle, pre-cal/radio specific caldata should be loaded first to get proper board id.
> 
> My understanding here is that there could be two different IPQ40xx based designs using the very same board.bin for one (or more) WiFi radios, as far as designs of their RF sides are the same (same parts and components values used)*.

It would have been much easier to deal with the IPQ4019 if Qualcomm went with the full-calibration data in ART just like they did with ath9k and early ath10k. 
As for why they did away with it: Well, the board-2.bin file can now be easily updated with a firmware update and I've seen this happen with the RT-AC58U and recently the A62 (and probably more).
There are also IPQ40XX where there are separate boardfiles for each regdomain (FCC, IC, EU).

Qualcomm had planned for this. In the pre-cal/board-2.bin there are fields which are reserved to [enumerate the customer and project id](https://www.mail-archive.com/ath10k@lists.infradead.org/msg06236.html). So the infrastructure was/is in place already.

> 
> Just btw, QCA9887 is Wave-2.

It is, it also uses the single board.bin template:
https://github.com/kvalo/ath10k-firmware/tree/master/QCA9887/hw1.0
https://source.codeaurora.org/quic/qsdk/oss/firmware/ath10k-firmware/plain/ath10k/QCA9887/hw1.0/

And the QCA9887 ART contains the full "cal" just like the QCA9880. (The MR33 has one, so I know about it from back then).

> > What I do find very odd in regards to how the board-2.bin for the QCA4019 is handled by QCA and their partners.
> 
> I don't.
> Different target markets, different internal teams taking care of the subject (?), plus... it's still Qualcomm: mess everywhere and with everything, wheels invented from scratch, datasheets for new ICs made with copy&paste approach with mistakes, etc.**

The ODM knew about it though (see Sven's customer,projectid post above). However, the design centers were left in the dark and this seems to 
cause the confusion.

> > Because for the customer QCA9377/QCA6174 device, Qualcomm takes much more care of the board-2.bin integration process. For example if you look into https://github.com/kvalo/ath10k-firmware/blob/master/QCA6174/hw3.0/board-2.bin you'll notice that there are many board definitions in there.
> 
> I think that market for QCA6174 is much bigger (phones, PCs running Windows, MiFi, etc.) and company focuses more resources on it than in case of IPQ40xx.

So, what can we do? Currently, I'm directing the people to the boardfile mailing-list threads in the hopes that they notice and ask their ODM what's going on. From what I can tell, this does seem to work. You can see this in this very PR (vs the old one #1655) . @zhangsd20 added "uses AP-DK07 BDF from QCA4019/hw1.0/board-2.bin" so there was some progress in this topic in the meantime.

Do you (or anyone?) see any avenues for improving this?
Because I would like IPQ40XX to be around QCA9377/QCA6174 in this regard and each device
will have its own boardfile so there can be no conflicts or hidden hiccups in the future.

> > They all have their unique identifiers (pciid + variant) and multiple boardfiles have the same checksum, so having a the data twice isn't a problem either.
> 
> Yeah but what's the purpose of having the very same binary data multiple times in that file? For me that's a resource wasting approach (why not assign single board.bin to several different pciid?) - anyone with QCA6174 card has to download almost 0.5 MB (for now, it will get bigger) binary just to make the card working.

This could be a great mailing-list topic... if Qualcomm was to reply to us/me/you :frowning_face: .
I don't know their "end-game" for the board-2.bin. For now the full IPQ4019 board-2.bin is 438KiB (
compressed with lzma(well, xz) -9 it goes to ~ 8KiB).

Now, if push comes to shove we still have the ipq-wifi package so the needed boardfiles for a device could be extracted the upstream board-2.bin library (thankfully, the official tool "ath10k-bdencoder" is open source and available in the https://github.com/qca/qca-swiss-army-knife repo) and individually packaged for just the device.  
 
regards,
Christian

**Comment by zhangsd20** [comment]
> Well, what is qcom,ath10k-calibration-variant = "Qxwlan-E2600AC"; used for if you are using upstream ap-dk07 board-2.bin?

I only modified the file in kernel4.14 and forgot to modify the file in kernel4.19. This is my fault, I am sorry.

**Comment by pepe2k** [comment]
@chunkeey 

> > > What I do find very odd in regards to how the board-2.bin for the QCA4019 is handled by QCA and their partners.
> > 
> > 
> > I don't.
> > Different target markets, different internal teams taking care of the subject (?), plus... it's still Qualcomm: mess everywhere and with everything, wheels invented from scratch, datasheets for new ICs made with copy&paste approach with mistakes, etc.**
> 
> The ODM knew about it though (see Sven's customer,projectid post above). However, the design centers were left in the dark and this seems to cause the confusion.

Maybe Qualcomm expected that this platform will be sold only over official channels and BDF ID assignment will be under their (or their certified distributors/design centers) control? We all know how it looks with cheap hardware designed and manufactured in Asia.

> > > Because for the customer QCA9377/QCA6174 device, Qualcomm takes much more care of the board-2.bin integration process. For example if you look into https://github.com/kvalo/ath10k-firmware/blob/master/QCA6174/hw3.0/board-2.bin you'll notice that there are many board definitions in there.
> > 
> > 
> > I think that market for QCA6174 is much bigger (phones, PCs running Windows, MiFi, etc.) and company focuses more resources on it than in case of IPQ40xx.
> 
> So, what can we do? Currently, I'm directing the people to the boardfile mailing-list threads in the hopes that they notice and ask their ODM what's going on. From what I can tell, this does seem to work. You can see this in this very PR (vs the old one #1655) . @zhangsd20 added "uses AP-DK07 BDF from QCA4019/hw1.0/board-2.bin" so there was some progress in this topic in the meantime.

My opinion here might be controversial. I think we should permanently drop `ipq-wifi` package and stop accepting support for devices which don't have BDF available in upstream. At the same time, we should include detailed description about this in our [Submitting patches](https://openwrt.org/submitting-patches) guide (it already consists great suggestions regarding DTS).

I really don't think we should deal with this mess in OpenWrt codebase. Currently, the same job is done twice - first, BDFs are packed and included in `ipq-wifi` and send to upstream... then, when files got included in upstream, we need to get rid of them from `ipq-wifi` and update device DTS. Doesn't make sense anymore, especially if now the upstreaming process of BDFs is known and widely tested.

> Do you (or anyone?) see any avenues for improving this?

As above, let upstream (and device support contributors) deal with the problem.

> Because I would like IPQ40XX to be around QCA9377/QCA6174 in this regard and each device
> will have its own boardfile so there can be no conflicts or hidden hiccups in the future.

See below:

> > > They all have their unique identifiers (pciid + variant) and multiple boardfiles have the same checksum, so having a the data twice isn't a problem either.
> > 
> > 
> > Yeah but what's the purpose of having the very same binary data multiple times in that file? For me that's a resource wasting approach (why not assign single board.bin to several different pciid?) - anyone with QCA6174 card has to download almost 0.5 MB (for now, it will get bigger) binary just to make the card working.
> 
> This could be a great mailing-list topic... if Qualcomm was to reply to us/me/you frowning_face .
> I don't know their "end-game" for the board-2.bin. For now the full IPQ4019 board-2.bin is 438KiB (
> compressed with lzma(well, xz) -9 it goes to ~ 8KiB).

> Now, if push comes to shove we still have the ipq-wifi package so the needed boardfiles for a device could be extracted the upstream board-2.bin library (thankfully, the official tool "ath10k-bdencoder" is open source and available in the https://github.com/qca/qca-swiss-army-knife repo) and individually packaged for just the device.

I'm worried about legal side of this. There should be a single, common source of these files and ath10k-firmware is the right place for it (from the beginning I considered `ipq-wifi` as a temporary workaround). Re-packing them during image generation might be the best way for us to do it without too much hassle (still thinking about legal side of this approach).

Cheers,
Piotr

**Comment by chunkeey** [comment]
> > > > What I do find very odd in regards to how the board-2.bin for the QCA4019 is handled by QCA and their partners.
> > > 
> > > I don't.
> > > Different target markets, different internal teams taking care of the subject (?), plus... it's still Qualcomm: mess everywhere and with everything, wheels invented from scratch, datasheets for new ICs made with copy&paste approach with mistakes, etc.**
> > 
> > 
> > The ODM knew about it though (see Sven's customer,projectid post above). However, the design centers were left in the dark and this seems to cause the confusion.
> 
> Maybe Qualcomm expected that this platform will be sold only over official channels and BDF ID assignment will be under their (or their certified distributors/design centers) control? We all know how it looks with cheap hardware designed and manufactured in Asia.

Well, I find some consolation in the fact that the "variant" hack was already implemented (broken, but @tchebb fixed it). They too had to deal with clashes on their QCA6174. so this made it easier for the IPQ4019 situation. Otherwise I don't think the "upstream boardfile" program would be a thing.

> > > > Because for the customer QCA9377/QCA6174 device, Qualcomm takes much more care of the board-2.bin integration process. For example if you look into https://github.com/kvalo/ath10k-firmware/blob/master/QCA6174/hw3.0/board-2.bin you'll notice that there are many board definitions in there.
> > > 
> > > 
> > > I think that market for QCA6174 is much bigger (phones, PCs running Windows, MiFi, etc.) and company focuses more resources on it than in case of IPQ40xx.
> > 
> > 
> > So, what can we do? Currently, I'm directing the people to the boardfile mailing-list threads in the hopes that they notice and ask their ODM what's going on. From what I can tell, this does seem to work. You can see this in this very PR (vs the old one #1655) . @zhangsd20 added "uses AP-DK07 BDF from QCA4019/hw1.0/board-2.bin" so there was some progress in this topic in the meantime.
> 
> My opinion here might be controversial. I think we should permanently drop `ipq-wifi` package and stop accepting support for devices which don't have BDF available in upstream. At the same time, we should include detailed description about this in our [Submitting patches](https://openwrt.org/submitting-patches) guide (it already consists great suggestions regarding DTS).
> 
> I really don't think we should deal with this mess in OpenWrt codebase. Currently, the same job is done twice - first, BDFs are packed and included in `ipq-wifi` and send to upstream... then, when files got included in upstream, we need to get rid of them from `ipq-wifi` and update device DTS. Doesn't make sense anymore, especially if now the upstreaming process of BDFs is known and widely tested.

Yeah, I'm not happy about it either since it takes months in order to get the boarddata upstream (In fact, you'll likely see similar results with the boardfile you submitted. Let's wait and watch).
But apart from that, I also posted pretty much the same "worries and thoughts" back in the day:  <https://www.mail-archive.com/ath10k@lists.infradead.org/msg06509.html>

> However, this is a bit of a problem for LEDE/OpenWRT.
>
>You see: The board-2.bin will have to be baked into the rootfs-image
>of the LEDE/OpenWRT firmware.
>But in order to do that, the board-2.bin needs to be extracted from
>the vendor's binary firmware image (via unsquashfs). Or if we are 
>lucky: from the source archive, if the vendor left them in there.
>
>The problem is: these raw board.bin files sourced from the binary
>firmware image lack a redistributable license.
>And therefore they can't be shipped with LEDE/OpenWRT. 
>(The raw board.bin files from the source archive are usually
>in the  madwifi-11n directory, which has a redistributable 
>license file and can be converted to board-2.bin with the 
>ath10k-bdencoder. But there's no explicit statement that
>this is possible.)
>
>The way I see it atm, developers would need to ask the vendors
>for  permission and LEDE/OpenWRT needs to setup something like
>a board-2.bin repository (similar to linux-firmware) for 
>all the IPQ40XX devices (and future devices?)...
>
>...
>
>Is there a way around this problem? 
>Can somebody ask around in the Legal Department? And clarify if:
>
>1. Whenever the board-2.bin are copyright-able in the first place?
>   INAL. These files just contain calibration values/parameters
>   and no programs. From what I know facts/numbers are not copyright-able.
>   But I don't know if it applies here or not. And if there is a
>   difference between US/Europe/China/... laws.
>
>2. What license these board.bin files should have.
>   They are generated by the halphy_tools' eepromUtil utility.
>   Are these files all original, or do they classify as
>   derived works?
>
>3. Most importantly: Can this be handled differently for future devices?
>   Like adding a redistributable license text inside the board.bin files?
>   (Like a header or trailer. A trailer would have the advantage that if
>   the caldata length is fixed (like 12064 bytes for IPQ40XX), the trailer
>   won't be uploaded to the device. So a trailer can be easily added to 
>   existing files).
>
>(4. Any workaround?
>    Like trying to get by with just the precal-data from flash like in [4]?)

And coincidentally this was the last time I ever heard from them :disappointed: .The ipq-wifi aka "board repository" is the surviving legacy and it was either that or not have the ipq40xx target.
 
Note on [4]: This board-2.bin issue also affects 9888 to some degree. You can see the w/a for it in ar71xx's [11-ath10k-caldata](https://github.com/openwrt/openwrt/blob/master/target/linux/ar71xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata#L176). Instead of dealing with the board.bin, they (re-)use the pre-cal-data there. Which is working OK in a pinch, but I was told that it is discouraged (but sadly not really: WHY it is discouraged.. Oh well).

> > Do you (or anyone?) see any avenues for improving this?
> 
> As above, let upstream (and device support contributors) deal with the problem.

I'm trying to help as best as I can, but I'm stumped about what to do in this case/PR. Because in the referenced #1655 you'll find that @zhangsd20 has indeed already sent a set of board files for the E2600AC-C1/C2 to the ath10k mailing list.

I took the chance and [I looked at the files and what I found was](https://github.com/openwrt/openwrt/pull/1655#issuecomment-452053480)

> I did take a peek into the board_wifi0_2g.bin, board_wifi0_5g.bin, board_wifi1_5g.bin from afar they seem to be 1:1 copies of the reference boards, but with the MAC address being changed. However the board checksum were not updated/recalculated?! This seems odd because from what I was told the QCA proprietary driver checks this and refuses incorrect checksums.

I'm still trying to figure out what was the purpose of this stunt? I tried talking to the committer about it, but for some reason nobody from Qxwlan is able or willing to answer on the PR and nobody sent a fixed file upstream either. That's why I'm very cautious here, because I do want to know what's going on and have a better understanding of the situation and topic than before.

> > > > They all have their unique identifiers (pciid + variant) and multiple boardfiles have the same checksum, so having a the data twice isn't a problem either.
> > > 
> > > 
> > > Yeah but what's the purpose of having the very same binary data multiple times in that file? For me that's a resource wasting approach (why not assign single board.bin to several different pciid?) - anyone with QCA6174 card has to download almost 0.5 MB (for now, it will get bigger) binary just to make the card working.
> > 
> > 
> > This could be a great mailing-list topic... if Qualcomm was to reply to us/me/you frowning_face .
> > I don't know their "end-game" for the board-2.bin. For now the full IPQ4019 board-2.bin is 438KiB (
> > compressed with lzma(well, xz) -9 it goes to ~ 8KiB).
> 
> > Now, if push comes to shove we still have the ipq-wifi package so the needed boardfiles for a device could be extracted the upstream board-2.bin library (thankfully, the official tool "ath10k-bdencoder" is open source and available in the https://github.com/qca/qca-swiss-army-knife repo) and individually packaged for just the device.
> 
> I'm worried about legal side of this. There should be a single, common source of these files and ath10k-firmware is the right place for it (from the beginning I considered `ipq-wifi` as a temporary workaround). Re-packing them during image generation might be the best way for us to do it without too much hassle (still thinking about legal side of this approach).

There is more than one legal side. Apart from "who owns these files", there's also the case that
without the "correct" file, the WiFi's will operate out of their parameters and violate regulatory compliance.

(no, the irony about the FCC restriction isn't lost on me here... since on IPQ4019
the situation seems to be: Unless you care your WiFi is operating out of spec.).

Regards,
Christian

**Comment by zhangsd20** [comment]
Hi，
I had no time to do the e2600ac because I was busy doing other things. I have re-sent the latest full board.bin to ath10k via email and ipq-wifi-pkg has been added. Please check it out, if you have any questions, please contact me in time.

I am very sorry about this.

**Comment by chunkeey** [comment]
Merged. Thanks.

**Comment by Neustradamus** [comment]
Merged commit:
- https://github.com/openwrt/openwrt/commit/bbab33724d5fbe3fad87d885adc16ed3dc47742f

Replaces:
- https://github.com/openwrt/openwrt/pull/1655

**Pull request closed without merging**